### PR TITLE
WebIDL Fix ImageData constructor to take a Uint8ClampedArray instead of js_object

### LINKED ISSUE #31320: WebIDL: Fix ImageData constructor to take a Uint8ClampedArray instead of object
https://github.com/servo/servo/blob/c367309a8fc004e47a6d8c0528b82e9d60a5716a/components/script/dom/webidls/CanvasRenderingContext2D.webidl#L257

This may require fiddling with `components/script/dom/bindings/codegen/CodegenRust.py`.

### PR DESCRIPTION
**Pull request by Taym95**
Fix ImageData constructor to take a Uint8ClampedArray instead of js_object

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #31320

- [x] These changes do not require tests because there is not behavior change.

### CODE REVIEW

**Comment by Taym95** [comment]
@gterzian Do you have any idea what to do with the PR? After this change, tests are failing for all browsers, including Servo?

**Comment by mrobinson** [comment]
@Taym95 What's the status of this PR?

**Comment by gterzian** [comment]
@Taym95 What is the failing test? If all browsers are failing, it may be because the test itself is not conformant.

**Comment by jdm** [comment]
https://wpt.fyi/results/html/semantics/embedded-content/the-canvas-element/imagedata.html?label=master&label=experimental&aligned&q=imagedata.html

File: components/script/dom/imagedata.rs [hunk 1]
```diff
@@ -33,42 +34,54 @@ pub struct ImageData {
...
     ) -> Fallible<DomRoot<ImageData>> {
-        let heap_typed_array = match new_initialized_heap_typed_array::<ClampedU8>(
-            HeapTypedArrayInit::Object(jsobject),
-        ) {
-            Ok(heap_typed_array) => heap_typed_array,
-            Err(_) => return Err(Error::JSFailed),
-        };
+        let cx = GlobalScope::get_cx();
+        let heap_typed_array =
+            match new_initialized_heap_typed_array::<ClampedU8>(HeapTypedArrayInit::Data {
```
Comment (gterzian) [inline components/script/dom/imagedata.rs:78]: I think this needs a rebase, because the api was updated.

File: components/script/dom/imagedata.rs [hunk 2]
```diff
@@ -33,42 +34,54 @@ pub struct ImageData {
...
+                data: unsafe { data.as_slice() },
+                cx: cx,
+            }) {
+                Ok(heap_typed_array) => heap_typed_array,
+                Err(_) => return Err(Error::JSFailed),
```
Comment (gterzian) [inline components/script/dom/imagedata.rs:83]: Either we are failing here where previously we wouldn't. Maybe try returning a type error here?

Or we are not producing a type error below at https://github.com/servo/servo/pull/31398/files#diff-74ee6b1aa263f228fb8648bb921e80b79654506e8b39570b764649a02f881402R89, don't know why. Maybe `get_buffer` is too liberal in returning an object even when it's of the wrong type?

**Review by gterzian — COMMENTED** [review]
I think we should find a way to make the test pass.

File: components/script_bindings/webidls/CanvasRenderingContext2D.webidl [hunk 3]
```diff
@@ -254,7 +254,7 @@ interface CanvasPattern {
  Serializable]
 interface ImageData {
   [Throws] constructor(unsigned long sw, unsigned long sh/*, optional ImageDataSettings settings = {}*/);
-  [Throws] constructor(/* Uint8ClampedArray */ object data, unsigned long sw, optional unsigned long sh
+  [Throws] constructor(Uint8ClampedArray data, unsigned long sw, optional unsigned long sh
```
[the remark this answers is not in the document]
Comment (jdm) [reply components/script_bindings/webidls/CanvasRenderingContext2D.webidl:257]: The specification is dense, but I think the test as written does not match the specification. From https://webidl.spec.whatwg.org/#js-overloads we can determine which overloaded constructor to call. The distinguishing arg is 0, so we start at step 12. 12.7 checks whether there is an overload with a typed array argument that matches the kind of the typed array that was passed in. There is not, so we continue to step 12.16, which checks if there is an overload with a number at the argument index (ignoring the actual value passed in at this point). There is, so that is the overload that is selected. We only passed this test previously because we reached step 12.11 and the argument was specified as an object, so we selected the overload with the typed array argument instead.

In short: the test is wrong, it is correct that we (and every other browser) fail it.

**Comment by mrobinson** [comment]
I don't even see a backtrace on this crash, which suggests it's a segfault and not a Rust panic:

```
  ▶ CRASH [expected OK] /css/compositing/canvas-composite-modes.html
  │ 
  └ Caught signal 11 in thread "Script(1,1)"
```

File: components/script/dom/bindings/buffer_source.rs [hunk 4]
```diff
@@ -63,36 +63,25 @@ pub(crate) enum BufferSource {
...
-            HeapBufferSource::<T>::new(BufferSource::ArrayBufferView(Heap::boxed(*array.handle())))
-        },
-    };
-    Ok(heap_buffer_source)
-}
+    rooted!(in (*cx) let mut array = ptr::null_mut::<JSObject>());
+    let typed_array_result =
+        create_buffer_source_with_length::<T>(cx, len as usize, array.handle_mut(), can_gc);
+    if typed_array_result.is_err() {
+        return Err(Error::JSFailed);
```
Comment (gterzian) [inline components/script/dom/bindings/buffer_source.rs:79]: Are you sure about the `Error::JSFailed`, previously it was just a `()`. Also the error is mapped to `JSFailed` a thttps://github.com/servo/servo/pull/31398/files#diff-74ee6b1aa263f228fb8648bb921e80b79654506e8b39570b764649a02f881402R61

File: components/script/dom/imagedata.rs [hunk 5]
```diff
@@ -55,31 +55,37 @@ impl ImageData {
...
+
                 let data = CreateWith::Slice(&d[..]);
                 Uint8ClampedArray::create(*cx, data, js_object.handle_mut()).unwrap();
-                Self::new_with_jsobject(global, None, width, Some(height), js_object.get(), can_gc)
+                Self::new_with_data(
+                    global,
+                    None,
+                    width,
+                    Some(height),
+                    CustomAutoRooterGuard::new(*cx, &mut CustomAutoRooter::new(typed_array)),
```
Comment (gterzian) [inline components/script/dom/imagedata.rs:70]: I think you want to assign this to a variable and then call the function. May be the source of the crash actually.

**Review by gterzian — CHANGES_REQUESTED** [review]
looks good overall, but we should find out where that crash is coming from.

File: components/script/dom/imagedata.rs [hunk 6]
```diff
@@ -55,31 +55,32 @@ impl ImageData {
...
+
+                let typed_array =
+                    create_buffer_source::<ClampedU8>(cx, &d[..], js_object.handle_mut(), can_gc)
+                        .map_err(|_| Error::JSFailed)?;
+
                 let data = CreateWith::Slice(&d[..]);
                 Uint8ClampedArray::create(*cx, data, js_object.handle_mut()).unwrap();
-                Self::new_with_jsobject(global, None, width, Some(height), js_object.get(), can_gc)
+                let mut rooted_typed_array = CustomAutoRooter::new(typed_array);
+                let data = CustomAutoRooterGuard::new(*cx, &mut rooted_typed_array);
```
Comment (jdm) [inline components/script/dom/imagedata.rs:66]: Use the `auto_root!` macro for this instead:
```rust
auto_root!(in(*cx) let data = typed_array);
```

**Review by gterzian — APPROVED** [review]
LGTM. Did we ascertain what caused the crash?

Also, let's track this as part of https://github.com/servo/servo/issues/30862
